### PR TITLE
feat: build quote tx

### PR DESCRIPTION
**Pull request by toshiSat**
## Description

Add Build/Execute Quote
Add Approval Steps
Fix minor bugs surrounding get quote ui

NOTE: Unchained isn't returning failed tx's or new tx's through websockets, this will need to be addressed.

Remaining Tickets: 
#267
#256 
lib#152

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?
- [x] Do all new and existing tests pass? Does the linter pass?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.

## Testing

Please outline all testing steps

1. Pull branch locally and run `yarn` to install new deps
2. etc...

### CODE REVIEW

File: src/assets/translations/messages-en.json
Comment (0xdef1cafe) [inline src/assets/translations/messages-en.json:62]: change this copy

File: src/components/Trade/TradeInput.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeInput.tsx:20]: change the lodash import to `'lodash/get'` above too please

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [inline src/components/Approval/Approval.tsx:30]: don't cast here

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [inline src/components/Approval/Approval.tsx:55]: don't cast here

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:37]: don't cast here

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/AssetToAsset.tsx:42]: stylistically i'd prefer a switch here but this is fine

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:104]: translations

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:117]: translations

File: src/components/Trade/TradeInput.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeInput.tsx:49]: give `useFormContext` a type here so we get typed values back

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [inline src/components/Approval/Approval.tsx:27]: give `useFormContext<SomeTypeHere>` a type

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [inline src/components/Approval/Approval.tsx:35]: give this a type

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:33]: give this a type

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [inline src/components/Approval/Approval.tsx:48]: stylistically early return pattern makes this less of a christmas tree and more readable

File: src/components/Trade/TradeInput.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeInput.tsx:56]: we can actually safely cast here to get types from `useWatch`

e.g.
```
useWatch({ listOfThings }) as [Form.Quote, Form.Action]
```

File: src/components/Trade/TradeInput.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeInput.tsx:63]: early return pattern

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [inline src/components/Approval/Approval.tsx:135]: this might fail

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (0xdef1cafe) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx:52]: also opt chain here

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:55]: early return

File: src/components/Trade/hooks/useSwapper/useSwapper.test.tsx
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/useSwapper.test.tsx:72]: needs another test for if it does need approval

File: src/pages/Assets/AssetDetails/AssetHistory.tsx
Comment (0xdef1cafe) [inline src/pages/Assets/AssetDetails/AssetHistory.tsx:41]: previous key was more robust

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yml
Comment (0xdef1cafe) [inline .github/workflows/pr.yml:12]: try deleting this and see if it still runs

File: src/pages/Assets/AssetDetails/AssetHistory.tsx
Comment (toshiSat) [reply src/pages/Assets/AssetDetails/AssetHistory.tsx:41]: This is causing an infinite loop

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (DaoDev44) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx:103]: This is just a label for the tool tip. Not user facing.

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (DaoDev44) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx:116]: same as above

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (DaoDev44) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx:124]: same as above

File: src/components/Approval/Approval.tsx
Comment (toshiSat) [reply src/components/Approval/Approval.tsx]: Good luck getting typescript to accept this

File: src/components/Approval/Approval.tsx
Comment (toshiSat) [inline src/components/Approval/Approval.tsx:2]: I disagree with this, now if someone needs to import another lodash function we need an entire like for it.  It seems counter intuitive to nitpick code this much

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx
Comment (toshiSat) [reply src/components/Trade/TradeConfirm/AssetToAsset.tsx:42]: This file will be deleted for components/AssetToAsset

File: src/components/Approval/Approval.tsx
Comment (cjthompson) [reply src/components/Approval/Approval.tsx:2]: `get` isn't even needed. You should be using `?.` instead of `get`.

File: src/components/Approval/Approval.tsx
Comment (cjthompson) [inline src/components/Approval/Approval.tsx:72]:
```suggestion
    const error = errors?.buildQuote?.message ?? null
```

File: src/components/Trade/TradeInput.tsx
Comment (cjthompson) [inline src/components/Trade/TradeInput.tsx:106]:
```suggestion
  const error = errors?.useSwapper?.message ?? null
```

File: src/components/Approval/Approval.tsx
Comment (DaoDev44) [inline src/components/Approval/Approval.tsx:53]: @0xdef1cafe still having issues with this one

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [reply src/components/Approval/Approval.tsx]: @toshiSat https://github.com/shapeshift/web/pull/264/commits/be0264f36b7a1c56af5a5fbdb8b1c3c99317baa1 🥱

File: src/components/Approval/Approval.tsx
Comment (0xdef1cafe) [reply src/components/Approval/Approval.tsx]: https://github.com/shapeshift/web/pull/264/commits/be0264f36b7a1c56af5a5fbdb8b1c3c99317baa1

**Comment by 0xdef1cafe** [comment]
LGTM

File: src/components/Trade/TradeInput.tsx
Comment (DaoDev44) [inline src/components/Trade/TradeInput.tsx:72]: @0xdef1cafe Make TradeInput not eth specific

**Review by 0xdef1cafe — APPROVED** [review]

File: src/components/Approval/Approval.tsx
Comment (cjthompson) [inline src/components/Approval/Approval.tsx:70]: No try/catch can result in an `UnhandlePromiseRejection` error

Comment (cjthompson) [reply src/components/Approval/Approval.tsx:70]: Someone else's PR includes a "poll" utility function that would simplify this code. You may want to go back a refactor after that gets merged.

File: src/components/Approval/Approval.tsx
Comment (cjthompson) [inline src/components/Approval/Approval.tsx:119]: I'm not thrilled with this being hard-coded inline. Config doesn't feel like the best place for it, but it might be better than here.At the very least it should be a static defined at the top of the file.

File: src/components/Approval/Approval.tsx
Comment (cjthompson) [inline src/components/Approval/Approval.tsx:134]: We really should have helper functions that create explorer links given a link template and a value.

File: src/components/Approval/Approval.tsx
Comment (cjthompson) [inline src/components/Approval/Approval.tsx:148]:
```suggestion
                  bnOrZero(fee)
```

File: src/components/Approval/Approval.tsx
Comment (cjthompson) [inline src/components/Approval/Approval.tsx:153]: hmm... toCrypto takes a Number and not a BigNumber? this could result in an incorrect value for very small amounts of ETH.

File: src/components/TokenRow/TokenRow.tsx
Comment (cjthompson) [inline src/components/TokenRow/TokenRow.tsx:27]: The convention for generics is to use `T` to represent the generic type, though it's not mandatory.
```suggestion
type TokenRowProps<T> = {
  control: Control<T>
  fieldName: Path<T>
```

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx
Comment (cjthompson) [inline src/components/Trade/TradeConfirm/AssetToAsset.tsx:42]: this is a nested ternary. A switch statement or a series of `if / else if` would be more readable. I don't mind nested ternaries for small one-liners but this is bigger and not as easy to parse.

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (cjthompson) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:56]: possible `UnhandledPromiseRejection` error

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (cjthompson) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:111]:
```suggestion
                    bnOrZero(quote?.rate || '')
```

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (cjthompson) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:124]:
```suggestion
                    bnOrZero(fees?.fee)
```

File: src/components/Trade/TradeInput.tsx
Comment (cjthompson) [inline src/components/Trade/TradeInput.tsx:66]:
```suggestion
    if (!(quote?.sellAsset && quote?.buyAsset && sellAsset.amount)) return
```

File: src/components/Trade/TradeInput.tsx
Comment (cjthompson) [inline src/components/Trade/TradeInput.tsx:160]: I assume `amount` and `sellAsset.amount` are strings? I'd prefer to use BigNumber to compare as it won't have precision errors.

File: src/components/Trade/TradeInput.tsx
Comment (cjthompson) [inline src/components/Trade/TradeInput.tsx:203]:
```suggestion
                } = ${firstNonZeroDecimal(bnOrZero(quote?.rate))} ${
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (cjthompson) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:150]: if you don't use `isFiat` later in the function
```suggestion
          if (action === TradeActions.FIAT) {
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (cjthompson) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:157]: What guarantee do you have that `rate` is a valid numeric value? Without a guard this chain can return `NaN` or throw a divide by zero error.
```suggestion
                bnOrZero(amount).div(rate).toString(),
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (cjthompson) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:187]: Make sure you are avoiding divide by zero errors and NaN values by guarding on the variables you expect to be valid numeric values.

```suggestion
            buyAssetUsdRate = bnOrZero(sellAssetUsdRate).div(newQuote?.rate || 1).toString()
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (cjthompson) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:195]: Or you can make `sellAssetUsdRate` / `buyAssetUsdRate` are BigNumbers at the top of the scope before they're used and avoid wrapping them in `bn` each time. You can then guard that it's a valid number once at the top.

```suggestion
          const sellAssetFiatRate = bnOrZerp(sellAssetUsdRate).times(1).toString() // TODO: Implement fiatPerUsd here
          const buyAssetFiatRate = bnOrZoer(buyAssetUsdRate).times(1).toString() // TODO: Implement fiatPerUsd here
```

**Review by elmutt — APPROVED** [review]

File: src/components/Approval/Approval.tsx
Comment (DaoDev44) [reply src/components/Approval/Approval.tsx:153]: I agree, but this is out of the scope of this pr.

File: src/components/Approval/Approval.tsx
Comment (toshiSat) [reply src/components/Approval/Approval.tsx]: and nothing of value was gained or lost 😂

File: src/pages/Assets/AssetDetails/AssetHistory.tsx
Comment (toshiSat) [reply src/pages/Assets/AssetDetails/AssetHistory.tsx]: I repeat, this is causing an infinite error message being thrown in the console

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (toshiSat) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:195]: Don't commit this, spelling errors.

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx
Comment (toshiSat) [reply src/components/Trade/TradeConfirm/AssetToAsset.tsx:42]: This file will be deleted

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (toshiSat) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx:56]: This is another ticket https://github.com/shapeshift/lib/issues/152

File: src/components/Approval/Approval.tsx
Comment (toshiSat) [reply src/components/Approval/Approval.tsx:70]: This is another ticket https://github.com/shapeshift/lib/issues/152

File: src/pages/Assets/AssetDetails/AssetHistory.tsx
Comment (DaoDev44) [reply src/pages/Assets/AssetDetails/AssetHistory.tsx]: Fixed the console issue by adding `key={0}` to the loader per their [docs](https://www.npmjs.com/package/react-infinite-scroller)

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (cjthompson) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx:104]: tooltips aren't user facing? isn't the tooltip what shows up when they hover over the ? icon?

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (cjthompson) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:258]:
```suggestion
            fromBaseUnit(bnOrZero(quote.buyAmount), buyAsset.currency.precision)
```

**Pull request merged**
